File: scripts/fix_silhouette_options.js
/**
 * Fix silhouette question options in Firestore via REST API.
 * Uses the Firebase CLI stored credentials.
 * 
 * Run with: node scripts/fix_silhouette_options.js
 */
const https = require('https');
const http = require('http');
const fs = require('fs');
const path = require('path');
const url = require('url');

const PROJECT_ID = 'geoquiz-7790d';
const DB_PATH = `projects/${PROJECT_ID}/databases/%28default%29`;
const QUESTIONS_PATH = `${DB_PATH}/documents/questions`;

// Helper: read Firebase CLI credentials
function readFirebaseCreds() {
  const home = process.env.USERPROFILE || process.env.HOME;
  const paths = [
    path.join(home, '.config', 'configstore', 'firebase-tools.json'),
    path.join(home, 'AppData', 'Roaming', 'configstore', 'firebase-tools.json'),
  ];
  for (const p of paths) {
    if (fs.existsSync(p)) {
      return JSON.parse(fs.readFileSync(p, 'utf-8'));
    }
  }
  return null;
}

// Helper: make HTTP request
function request(method, urlStr, body, token) {
  return new Promise((resolve, reject) => {
    const u = new URL(urlStr);
    const options = {
      hostname: u.hostname,
      port: u.port || 443,
      path: u.pathname + u.search,
      method,
      headers: {
        'Authorization': `Bearer ${token}`,
        'Content-Type': 'application/json',
      },
    };
    if (body) options.headers['Content-Length'] = Buffer.byteLength(body);

    const req = https.request(options, res => {
      let data = '';
      res.on('data', chunk => data += chunk);
      res.on('end', () => {
        if (res.statusCode >= 200 && res.statusCode < 300) {
          try { resolve(JSON.parse(data)); }
          catch { resolve(data); }
        } else {
          reject(new Error(`HTTP ${res.statusCode}: ${data.slice(0, 200)}`));
        }
      });
    });
    req.on('error', reject);
    if (body) req.write(body);
    req.end();
  });
}

// Helper: format value for Firestore REST API
function toFieldValue(v) {
  if (typeof v === 'string') return { stringValue: v };
  if (typeof v === 'boolean') return { booleanValue: v };
  if (typeof v === 'number') {
    if (Number.isInteger(v)) return { integerValue: String(v) };
    return { doubleValue: v };
  }
  if (Array.isArray(v)) {
    return { arrayValue: { values: v.map(toFieldValue) } };
  }
  if (v === null || v === undefined) return { nullValue: 'NULL_VALUE' };
  if (typeof v === 'object') {
    const fields = {};
    for (const [k, val] of Object.entries(v)) {
      fields[k] = toFieldValue(val);
    }
    return { mapValue: { fields } };
  }
  return { stringValue: String(v) };
}

// Helper: unwrap Firestore field value
function fromFieldValue(v) {
  if (v.stringValue !== undefined) return v.stringValue;
  if (v.integerValue !== undefined) return parseInt(v.integerValue);
  if (v.doubleValue !== undefined) return v.doubleValue;
  if (v.booleanValue !== undefined) return v.booleanValue;
  if (v.arrayValue) {
    return (v.arrayValue.values || []).map(fromFieldValue);
  }
  if (v.mapValue) {
    const obj = {};
    for (const [k, val] of Object.entries(v.mapValue.fields || {})) {
      obj[k] = fromFieldValue(val);
    }
    return obj;
  }
  return null;
}

// Refresh Firebase token
function refreshToken(creds) {
  return new Promise((resolve, reject) => {
    const refreshToken = creds.tokens.refresh_token;
    if (!refreshToken) return reject(new Error('No refresh token'));

    const data = new url.URLSearchParams({
      grant_type: 'refresh_token',
      refresh_token: refreshToken,
      client_id: '563584335869-fgrhgmd47bqnekij5i8b5pr03ho849e6.apps.googleusercontent.com',
    }).toString();

    const req = https.request({
      hostname: 'oauth2.googleapis.com',
      path: '/token',
      method: 'POST',
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
        'Content-Length': Buffer.byteLength(data),
      },
    }, res => {
      let body = '';
      res.on('data', chunk => body += chunk);
      res.on('end', () => {
        if (res.statusCode === 200) {
          const result = JSON.parse(body);
          // Update stored credentials
          creds.tokens.access_token = result.access_token;
          creds.tokens.expires_at = Date.now() + (result.expires_in || 3600) * 1000;
          const home = process.env.USERPROFILE || process.env.HOME;
          const p = path.join(home, '.config', 'configstore', 'firebase-tools.json');
          fs.writeFileSync(p, JSON.stringify(creds, null, 2));
          resolve(result.access_token);
        } else {
          reject(new Error(`Refresh failed: ${res.statusCode} - ${body}`));
        }
      });
    });
    req.on('error', reject);
    req.write(data);
    req.end();
  });
}

async function main() {
  console.log('='.repeat(60));
  console.log('  GEO C - Fix Silhouette Options');
  console.log('='.repeat(60));

  // Step 1: Get Firebase token
  console.log('\n1. Getting Firebase token...');
  const creds = readFirebaseCreds();
  if (!creds) {
    console.error('ERROR: Firebase credentials not found. Run firebase login first.');
    process.exit(1);
  }

  let token = creds.tokens.access_token;
  const expiresAt = creds.tokens.expires_at || 0;
  const now = Date.now();
  
  if (expiresAt > now + 60000) {
    console.log('   Token still valid');
  } else {
    console.log('   Token expired, refreshing...');
    try {
      token = await refreshToken(creds);
      console.log('   Token refreshed!');
    } catch (e) {
      console.error('   Token refresh failed:', e.message);
      console.log('   Trying with stored token (may fail)...');
    }
  }

  // Step 2: Extract country names from questions
  console.log('\n2. Extracting country names...');
  const allCountries = new Set();
  
  const queryBody = JSON.stringify({
    structuredQuery: {
      from: [{ collectionId: 'questions' }],
      where: {
        compositeFilter: {
          op: 'OR',
          filters: [
            { fieldFilter: { field: { fieldPath: 'type' }, op: 'EQUAL', value: { stringValue: 'flag' } } },
            { fieldFilter: { field: { fieldPath: 'type' }, op: 'EQUAL', value: { stringValue: 'silhouette' } } },
          ],
        },
      },
      select: { fields: [{ fieldPath: 'correctAnswer' }] },
      limit: 500,
    },
  });

  const queryUrl = `https://firestore.googleapis.com/v1/${DB_PATH}/documents:runQuery`;
  const queryResult = await request('POST', queryUrl, queryBody, token);
  
  queryResult.forEach(item => {
    if (item.document && item.document.fields) {
      const answer = fromFieldValue(item.document.fields.correctAnswer);
      if (answer && answer.length > 2) allCountries.add(answer);
    }
  });
  
  const countries = [...allCountries].sort();
  console.log(`   Found ${countries.length} country names`);

  // Step 3: Query silhouettes from Firestore
  console.log('\n3. Fetching silhouettes from Firestore...');
  
  const silQueryBody = JSON.stringify({
    structuredQuery: {
      from: [{ collectionId: 'questions' }],
      where: {
        fieldFilter: {
          field: { fieldPath: 'type' },
          op: 'EQUAL',
          value: { stringValue: 'silhouette' },
        },
      },
      limit: 500,
    },
  });

  const silResult = await request('POST', queryUrl, silQueryBody, token);
  const silhouettes = [];
  
  silResult.forEach(item => {
    if (item.document) {
      const doc = item.document;
      const docId = doc.name.split('/').pop();
      const fields = {};
      for (const [k, v] of Object.entries(doc.fields || {})) {
        fields[k] = fromFieldValue(v);
      }
      silhouettes.push({ id: docId, ...fields });
    }
  });
  
  console.log(`   Found ${silhouettes.length} silhouettes`);

  if (silhouettes.length === 0) {
    console.log('   No silhouettes found!');
    process.exit(0);
  }

  // Step 4: Show sample before
  console.log('\n4. Sample before:');
  silhouettes.slice(0, 3).forEach(s => {
    console.log(`   ${s.id}: correct=${s.correctAnswer}, options=${JSON.stringify(s.options)}`);
  });

  // Step 5: Fix options
  console.log('\n5. Fixing options...');
  
  const countrySet = new Set(countries);
  let nonsenseBefore = 0;
  
  silhouettes.forEach(s => {
    const opts = s.options || [];
    const bad = opts.filter(o => o !== s.correctAnswer && !countrySet.has(o));
    if (bad.length > 0) nonsenseBefore++;
  });
  console.log(`   Nonsense options before: ${nonsenseBefore}/${silhouettes.length}`);

  // Generate new options for each silhouette
  const updates = silhouettes.map(s => {
    const correct = s.correctAnswer;
    if (!correct) return null;

    const available = countries.filter(c => c !== correct);
    if (available.length < 3) {
      console.log(`   WARNING: Not enough countries for ${correct}`);
      return null;
    }

    // Pick 3 random other countries
    const shuffled = [...available].sort(() => Math.random() - 0.5);
    const distractors = shuffled.slice(0, 3);
    const newOptions = [...distractors, correct].sort(() => Math.random() - 0.5);

    return { id: s.id, correctAnswer: correct, options: newOptions };
  }).filter(Boolean);

  console.log(`   Will update ${updates.length} silhouettes`);

  // Step 6: Upload to Firestore using batchWrite
  console.log('\n6. Uploading to Firestore...');
  
  const BATCH_SIZE = 100;
  let uploaded = 0;
  
  for (let i = 0; i < updates.length; i += BATCH_SIZE) {
    const batch = updates.slice(i, i + BATCH_SIZE);
    const writes = batch.map(s => ({
      update: {
        name: `${QUESTIONS_PATH}/${s.id}`,
        fields: {
          options: toFieldValue(s.options),
          correctAnswer: toFieldValue(s.correctAnswer),
        },
      },
    }));

    const batchUrl = `https://firestore.googleapis.com/v1/${DB_PATH}/documents:batchWrite`;
    const result = await request('POST', batchUrl, JSON.stringify({ writes }), token);
    uploaded += batch.length;
    const pct = ((i + batch.length) / updates.length * 100).toFixed(0);
    console.log(`   Batch ${Math.floor(i / BATCH_SIZE) + 1}/${Math.ceil(updates.length / BATCH_SIZE)} (${pct}%): ${batch.length} updated (total: ${uploaded}/${updates.length})`);
  }

  console.log(`\n7. ✅ Done! ${uploaded}/${updates.length} silhouettes updated.`);

  // Step 8: Verify
  console.log('\n8. Verifying...');
  const verifyResult = await request('POST', queryUrl, silQueryBody, token);
  const verifySilhouettes = [];
  
  verifyResult.forEach(item => {
    if (item.document) {
      const doc = item.document;
      const docId = doc.name.split('/').pop();
      const fields = {};
      for (const [k, v] of Object.entries(doc.fields || {})) {
        fields[k] = fromFieldValue(v);
      }
      verifySilhouettes.push({ id: docId, ...fields });
    }
  });

  let good = 0;
  verifySilhouettes.forEach(s => {
    const opts = s.options || [];
    if (opts.length >= 4) {
      const allValid = opts.every(opt => countrySet.has(opt));
      if (allValid) good++;
    }
  });
  console.log(`   Silhouettes with valid options: ${good}/${verifySilhouettes.length}`);
  
  console.log('\n   Final sample:');
  verifySilhouettes.slice(0, 5).forEach(s => {
    console.log(`   ${s.id}: options=${JSON.stringify(s.options)}`);
  });

  console.log('\n✅ All done!');
}

main().catch(e => {
  console.error('\nFATAL:', e.message);
  process.exit(1);
});

File: scripts/fix_silhouette_options.py
"""
Corregir opciones de preguntas de silueta en Firestore.
Reemplaza las opciones sin sentido (monumentos, monedas, idiomas, etc.)
solo por nombres de paises como distractores.
"""
import json
import random
import urllib.request
import urllib.parse
import time
import os
import pathlib
import sys

PROJECT_ID = "geoquiz-7790d"
DB_PATH = f"projects/{PROJECT_ID}/databases/%28default%29/documents/questions"
BASE_URL = f"https://firestore.googleapis.com/v1/{DB_PATH}"

def get_access_token():
    cred_path = pathlib.Path.home() / ".config" / "configstore" / "firebase-tools.json"
    if not cred_path.exists():
        cred_path = pathlib.Path.home() / "AppData" / "Roaming" / "configstore" / "firebase-tools.json"
    if not cred_path.exists():
        print("No Firebase credentials found.")
        return None
    
    with open(cred_path, "r", encoding="utf-8") as f:
        data = json.load(f)
    
    tokens = data.get("tokens", {})
    access_token = tokens.get("access_token")
    refresh_token = tokens.get("refresh_token")
    
    expires_at = tokens.get("expires_at", 0)
    if expires_at and expires_at > time.time() + 60:
        print(f"Token valid (expires in {int(expires_at - time.time())}s)")
        return access_token
    
    # Try multiple known client secrets for Firebase CLI OAuth client
    client_secrets = [
        "j9iVZfS8kk8upbheD6mF2Vap",
        "j9iVZfS8kk8fyFhMU95pPqB7",
    ]
    if refresh_token:
        CLIENT_ID = "563584335869-fgrhgmd47bqnekij5i8b5pr3ho849se6.apps.googleusercontent.com"
        for secret in client_secrets:
            refresh_data = urllib.parse.urlencode({
                "grant_type": "refresh_token",
                "refresh_token": refresh_token,
                "client_id": CLIENT_ID,
                "client_secret": secret,
            }).encode("utf-8")
            req = urllib.request.Request(
                "https://oauth2.googleapis.com/token",
                data=refresh_data,
                headers={"Content-Type": "application/x-www-form-urlencoded"},
                method="POST"
            )
            try:
                with urllib.request.urlopen(req, timeout=30) as resp:
                    new_tokens = json.loads(resp.read().decode())
                    access_token = new_tokens["access_token"]
                    tokens["access_token"] = access_token
                    tokens["expires_at"] = time.time() + new_tokens.get("expires_in", 3600)
                    data["tokens"] = tokens
                    with open(cred_path, "w", encoding="utf-8") as f:
                        json.dump(data, f)
                    print("Token refreshed!")
                    return access_token
            except Exception as e:
                print(f"Token refresh failed with secret ending ...{secret[-4:]}: {e}")
    
    print("WARNING: Using stored token (may be expired)")
    return access_token

def field(v):
    if isinstance(v, str): return {"stringValue": v}
    elif isinstance(v, bool): return {"booleanValue": v}
    elif isinstance(v, int): return {"integerValue": str(v)}
    elif isinstance(v, float): return {"doubleValue": v}
    elif isinstance(v, list): return {"arrayValue": {"values": [field(item) for item in v]}}
    elif isinstance(v, dict): return {"mapValue": {"fields": {k: field(val) for k, val in v.items()}}}
    elif v is None: return {"nullValue": "NULL_VALUE"}
    return {"stringValue": str(v)}

def get_all_country_names():
    """Extract all country names from flag questions in the merged data."""
    countries = set()
    files_to_check = [
        "questions_fixed_options.json",
        "questions_all_merged.json",
        "questions_clean.json",
    ]
    script_dir = os.path.dirname(os.path.abspath(__file__))
    
    for fname in files_to_check:
        fpath = os.path.join(script_dir, fname)
        if not os.path.exists(fpath):
            continue
        with open(fpath, "r", encoding="utf-8") as f:
            try:
                questions = json.load(f)
                for q in questions:
                    if q.get("type") in ("flag", "silhouette"):
                        ans = q.get("correctAnswer", "")
                        if ans and len(ans) > 2:
                            countries.add(ans)
            except:
                pass
    
    return sorted(countries)

def fix_options_for_silhouette(question, all_countries):
    """Generate proper options for a silhouette question."""
    correct = question.get("correctAnswer", "")
    if not correct:
        return question
    
    available = [c for c in all_countries if c != correct]
    if len(available) < 3:
        question["options"] = []
        return question
    
    distractors = random.sample(available, 3)
    options = distractors + [correct]
    random.shuffle(options)
    question["options"] = options
    return question

def query_silhouettes_from_firestore(access_token):
    """Get all silhouette questions from Firestore."""
    query_url = f"{BASE_URL}:runQuery"
    query_body = json.dumps({
        "structuredQuery": {
            "from": [{"collectionId": "questions"}],
            "where": {
                "fieldFilter": {
                    "field": {"fieldPath": "type"},
                    "op": "EQUAL",
                    "value": {"stringValue": "silhouette"}
                }
            },
            "limit": 500
        }
    }).encode("utf-8")
    
    req = urllib.request.Request(query_url, data=query_body, method="POST")
    req.add_header("Authorization", f"Bearer {access_token}")
    req.add_header("Content-Type", "application/json")
    
    try:
        with urllib.request.urlopen(req, timeout=30) as resp:
            results = json.loads(resp.read().decode())
        silhouettes = []
        for item in results:
            if "document" not in item:
                continue
            doc = item["document"]
            fields_raw = doc.get("fields", {})
            name = doc["name"]
            doc_id = name.split("/")[-1]
            
            # Convert Firestore fields back to dict
            def unwrap(v):
                if "stringValue" in v: return v["stringValue"]
                if "integerValue" in v: return int(v["integerValue"])
                if "booleanValue" in v: return v["booleanValue"]
                if "arrayValue" in v:
                    return [unwrap(x) for x in v["arrayValue"].get("values", [])]
                if "mapValue" in v:
                    return {k: unwrap(val) for k, val in v["mapValue"].get("fields", {}).items()}
                return None
            
            q = {"id": doc_id}
            for k, v in fields_raw.items():
                q[k] = unwrap(v)
            silhouettes.append(q)
        return silhouettes
    except Exception as e:
        print(f"Error querying Firestore: {e}")
        return None

def upload_fixes_to_firestore(fixed_questions, access_token):
    """Upload fixed silhouette questions to Firestore using batchWrite."""
    BATCH_SIZE = 100
    total = len(fixed_questions)
    uploaded = 0
    
    for i in range(0, total, BATCH_SIZE):
        batch = fixed_questions[i:i+BATCH_SIZE]
        writes = []
        for q in batch:
            doc_id = q["id"]
            # Only update: options, extraData
            fields = {"options": field(q.get("options", [])),
                      "extraData": field(q.get("extraData", {})),
                      "imageUrl": field(q.get("imageUrl", "")),
                      "correctAnswer": field(q.get("correctAnswer", ""))}
            writes.append({"update": {"name": f"{DB_PATH}/{doc_id}", "fields": fields}})
        
        body = json.dumps({"writes": writes}).encode("utf-8")
        url = f"https://firestore.googleapis.com/v1/projects/{PROJECT_ID}/databases/%28default%29/documents:batchWrite"
        headers = {"Content-Type": "application/json", "Authorization": f"Bearer {access_token}"}
        req = urllib.request.Request(url, data=body, headers=headers, method="POST")
        
        try:
            with urllib.request.urlopen(req, timeout=60) as resp:
                uploaded += len(batch)
                pct = (i + len(batch)) / total * 100
                print(f"  Batch {i//BATCH_SIZE + 1}/{(total-1)//BATCH_SIZE + 1} ({pct:.0f}%): {len(batch)} uploaded (total: {uploaded}/{total})")
        except Exception as e:
            print(f"  Error in batch {i//BATCH_SIZE + 1}: {e}")
        
        time.sleep(0.5)
    
    return uploaded

def main():
    script_dir = os.path.dirname(os.path.abspath(__file__))
    os.chdir(script_dir)
    
    print("=" * 60)
    print("  GEO C - Fix Silhouette Options")
    print("=" * 60)
    
    # Step 1: Get access token
    print("\n1. Getting Firebase access token...")
    token = get_access_token()
    if not token:
        print("ERROR: No Firebase token")
        sys.exit(1)
    print("   OK")
    
    # Step 2: Get all country names from question data
    print("\n2. Extracting country names from questions...")
    all_countries = get_all_country_names()
    print(f"   Found {len(all_countries)} country names")
    
    # Step 3: Query Firestore for current silhouettes
    print("\n3. Fetching silhouettes from Firestore...")
    silhouettes = query_silhouettes_from_firestore(token)
    if silhouettes is None:
        print("   Could not query Firestore. Reading from local files...")
        sil_file = os.path.join(script_dir, "questions_silhouettes.json")
        with open(sil_file, "r", encoding="utf-8") as f:
            silhouettes = json.load(f)
        print(f"   Loaded {len(silhouettes)} from local file")
    else:
        print(f"   Found {len(silhouettes)} silhouettes in Firestore")
    
    if not silhouettes:
        print("ERROR: No silhouettes found")
        sys.exit(1)
    
    # Step 4: Show sample before
    print("\n4. Sample before fix:")
    for s in silhouettes[:3]:
        opts = s.get("options", [])
        print(f"   {s['id']}: correct={s.get('correctAnswer','?')}, options={opts}")
    
    # Step 5: Fix options
    print(f"\n5. Fixing options for {len(silhouettes)} silhouettes...")
    fix_count = 0
    for q in silhouettes:
        old_opts = q.get("options", [])
        q = fix_options_for_silhouette(q, all_countries)
        new_opts = q.get("options", [])
        if old_opts != new_opts:
            fix_count += 1
    print(f"   Fixed {fix_count} silhouettes")
    
    # Step 6: Show sample after
    print("\n6. Sample after fix:")
    for s in silhouettes[:3]:
        opts = s.get("options", [])
        print(f"   {s['id']}: correct={s.get('correctAnswer','?')}, options={opts}")
    
    # Step 7: Confirm and upload
    print(f"\n7. Uploading {len(silhouettes)} fixed silhouettes to Firestore...")
    
    uploaded = upload_fixes_to_firestore(silhouettes, token)
    print(f"\n8. Done! {uploaded}/{len(silhouettes)} silhouettes updated in Firestore.")
    
    # Step 8: Verify
    print("\n9. Verifying...")
    verify = query_silhouettes_from_firestore(token)
    if verify:
        good = sum(1 for s in verify if len(s.get("options", [])) == 4)
        print(f"   Silhouettes with 4 options: {good}/{len(verify)}")
        for s in verify[:3]:
            print(f"   {s['id']}: options={s.get('options', [])}")
    
    print("\nDone!")

if __name__ == "__main__":
    main()
